src/metadata/trackmetadatataglib.cpp

```diff
@@ -0,0 +1,688 @@
+    TagLib::ID3v2::FrameList covertArtFrame = tag.frameListMap()["APIC"];
+    if (!covertArtFrame.isEmpty()) {
+        TagLib::ID3v2::AttachedPictureFrame* picframe =
+                static_cast<TagLib::ID3v2::AttachedPictureFrame*>(covertArtFrame.front());
+        TagLib::ByteVector data = picframe->picture();
+        coverArt = QImage::fromData(
+                reinterpret_cast<const uchar *>(data.data()), data.size());
+    }
+    return coverArt;
+}
+
+QImage readAPETagCover(const TagLib::APE::Tag& tag) {
+    QImage coverArt;
+    if (tag.itemListMap().contains("COVER ART (FRONT)")) {
+        const TagLib::ByteVector nullStringTerminator(1, 0);
+        TagLib::ByteVector item =
+                tag.itemListMap()["COVER ART (FRONT)"].value();
+        int pos = item.find(nullStringTerminator);  // skip the filename
+        if (++pos > 0) {
+            const TagLib::ByteVector& data = item.mid(pos);
+            coverArt = QImage::fromData(
+                    reinterpret_cast<const uchar *>(data.data()), data.size());
+        }
+    }
+    return coverArt;
+}
+
+QImage readXiphCommentCover(const TagLib::Ogg::XiphComment& tag) {
+    QImage coverArt;
+    if (tag.fieldListMap().contains("METADATA_BLOCK_PICTURE")) {
+        QByteArray data(
+                QByteArray::fromBase64(
+                        tag.fieldListMap()["METADATA_BLOCK_PICTURE"].front().toCString()));
+        TagLib::ByteVector tdata(data.data(), data.size());
+        TagLib::FLAC::Picture p(tdata);
+        data = QByteArray(p.data().data(), p.data().size());
+        coverArt = QImage::fromData(data);
+    } else if (tag.fieldListMap().contains("COVERART")) {
+        QByteArray data(
+                QByteArray::fromBase64(
+                        tag.fieldListMap()["COVERART"].toString().toCString()));
+        coverArt = QImage::fromData(data);
+    }
+    return coverArt;
+}
+
+QImage readMP4TagCover(/*const*/TagLib::MP4::Tag& tag) {
+    QImage coverArt;
+    if (tag.itemListMap().contains("covr")) {
+        TagLib::MP4::CoverArtList coverArtList =
+                tag.itemListMap()["covr"].toCoverArtList();
+        TagLib::ByteVector data = coverArtList.front().data();
+        coverArt = QImage::fromData(
+                reinterpret_cast<const uchar *>(data.data()), data.size());
+    }
+    return coverArt;
+}
+
+namespace {
+
+void replaceID3v2Frame(TagLib::ID3v2::Tag* pTag, TagLib::ID3v2::Frame* pFrame) {
+    pTag->removeFrames(pFrame->frameID());
+    pTag->addFrame(pFrame);
+}
+
+void writeID3v2TextIdentificationFrame(TagLib::ID3v2::Tag* pTag,
+        const TagLib::ByteVector &id, const QString& text) {
+    TagLib::String::Type textType;
+    QByteArray textData;
+    if (4 <= pTag->header()->majorVersion()) {
+        // prefer UTF-8 for ID3v2.4.0 or higher
+        textType = TagLib::String::UTF8;
+        textData = text.toUtf8();
+    } else {
+        textType = TagLib::String::Latin1;
+        textData = text.toLatin1();
```

Comment:
What is the story for non latin1 characters? Should we check this and offer a workaround? 
For now a debug message would be nice.